File: Difficulty/Hard/[45] Jump Game II/solution.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        step, cur_index= 0, 0
        while cur_index != len(nums) - 1:
            if cur_index + nums[cur_index] >= len(nums) - 1:
                return step + 1
            max_shift = 0
            sub_optimal_index = 0
            for i in range(1, nums[cur_index] + 1):
                if i + nums[i+cur_index] > max_shift:
                    max_shift = i + nums[i + cur_index]
                    sub_optimal_index = cur_index + i
            cur_index = sub_optimal_index
            step += 1
        return step    

File: Difficulty/Hard/[45] Jump Game II/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int step = 0, cur_index = 0;
        while(cur_index != nums.size() - 1){
            if (cur_index + nums[cur_index] >= nums.size() - 1)
                return step + 1;
            int max_shift = 0;
            int sub_optimal_index = 0;
            for (int i = 1; i <= nums[cur_index]; ++i){
                if ((i + nums[i + cur_index]) > max_shift){
                    max_shift = i + nums[i + cur_index];
                    sub_optimal_index = cur_index + i;
                }
            }
            cur_index = sub_optimal_index;
            step++;
        }
        return step;
    }
};